File: solutions/110-balanced-binary-tree/balanced-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        def dfs(node):
            if not node:
                return 0
            leftDepth = dfs(node.left)
            rightDepth = dfs(node.right)
            if leftDepth == -1 or rightDepth == -1:
                return -1
            if abs(leftDepth - rightDepth) > 1:
                return -1
            return max(leftDepth, rightDepth) + 1
        isBalanced = dfs(root)
        if isBalanced == -1: return False
        return True

File: solutions/110-balanced-binary-tree/balanced-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean isBalanced(TreeNode root) {
        if(root == null){
            return true;
        }
        int a = checkH(root.left);
        int b = checkH(root.right);
        if(Math.abs(a-b) > 1){
            return false;
        }

        return isBalanced(root.left) && isBalanced(root.right);

    }

    public int checkH(TreeNode root){
        if(root == null){
            return 0;
        }

        return 1 + Math.max(checkH(root.left), checkH(root.right));
    }
}